clang/lib/CConv/RewriteUtils.cpp

```diff
@@ -1322,10 +1338,51 @@ class CastPlacementVisitor :
       Writer.InsertTextBefore(E->getBeginLoc(), CastPrefix);
     }
   }
+
+  bool needsFancyCast(std::set<ConstraintVariable *> &LHSCons, Expr *E,
+                      std::string &CastStr) {
+    ConstraintVariable *LCVariable = nullptr;
+    bool LHSChkType = false;
+    bool NeedFancyCast = false;
+    auto &CS = Info.getConstraints();
+    for (auto *LC : LHSCons) {
+      if (LC->anyChanges(CS.getVariables())) {
+        LCVariable = LC;
+        LHSChkType = true;
+        break;
+      }
+    }
+    if (LHSChkType) {
+      assert(LCVariable != nullptr && "Expected non-null");
+
+      // Now, check if RHS is either explicit cast or addr-of (&) expression
+      // in which case we insert fancy cast.
+      E = E->IgnoreParenImpCasts();
+      if (!isNULLExpression(E, *Context) &&
+          dyn_cast<ExplicitCastExpr>(E)) {
+        NeedFancyCast = true;
+      }
+      if (UnaryOperator *UO = dyn_cast<UnaryOperator>(E)) {
+        if (UO->getOpcode() == UO_AddrOf) {
+          NeedFancyCast = true;
+        }
+      }
+
+      if (NeedFancyCast) {
+        CastStr = "_Assume_bounds_cast<" +
+                  LCVariable->mkString(CS.getVariables(), false) +
+                  ">(";
+      }
+    }
+
```

Comment:
I'm seeing a bunch of places where this `_Assume_bounds_cast` is getting added to code I'm testing, but I don't see why it's being added. @jackastner just added an issue about this: https://github.com/plum-umd/checkedc-clang/issues/112

For the case you initially added it for, in `vsftpd`, I think it's not the right fix anyway. The problem is that we are not handling `&(p->x[i])` expressions properly. But we are about to do that (fixing issue https://github.com/plum-umd/checkedc-clang/issues/106).

In sum: Maybe we should comment out the stuff involving `_Assume_bounds_cast` ?

Reply:
Yeah, I'm wondering if this code can drop. It's weird that it's looking specifically for `&`, which I don't know why it should? What's the justification for that?